core/platform/auth/firebase_auth_services.py

```diff
@@ -393,3 +381,107 @@ def associate_multi_auth_ids_with_user_ids(auth_id_user_id_pairs):
         auth_models.UserAuthDetailsModel.update_timestamps_multi(
             assoc_by_user_id_models)
         auth_models.UserAuthDetailsModel.put_multi(assoc_by_user_id_models)
+
+
+@contextlib.contextmanager
+def _firebase_admin_context():
+    """Returns a context for calling the Firebase Admin SDK.
+
+    Yields:
+        None. No relevent context expression.
+    """
+    # NOTE: "app" is the term Firebase uses for the "entry point" to the
+    # Firebase SDK. Oppia only has one server, so it only needs to instantiate
+    # one app.
+    firebase_connection = firebase_admin.initialize_app(
+        options={'projectId': feconf.OPPIA_PROJECT_ID})
+    try:
+        yield
+    finally:
+        # NOTE: This is not dangerous. We are just deleting the resources used
+        # to form a connection to Firebase servers.
+        firebase_admin.delete_app(firebase_connection)
+
+
+def _get_session_cookie(request):
+    """Returns the session cookie authorizing the signed in user, if present.
+
+    Args:
+        request: webapp2.Request. The HTTP request to inspect.
+
+    Returns:
+        str|None. Value of the session cookie authorizing the signed in user, if
+        present, otherwise None.
+    """
+    return request.cookies.get(feconf.FIREBASE_SESSION_COOKIE_NAME)
+
+
+def _get_id_token(request):
+    """Returns the ID token authorizing a user, or None if missing.
+
+    Oppia uses the OAuth 2.0's Bearer authentication scheme to send ID Tokens.
+
+    Bearer authentication (a.k.a. token authentication) is an HTTP
+    authentication scheme based on "bearer tokens", an encrypted JWT generated
+    by a trusted identity provider in response to login requests.
+
+    The name "Bearer authentication" can be understood as: "give access to the
+    bearer of this token." These tokens _must_ be sent in the `Authorization`
+    header of HTTP requests, and _must_ have the format: `Bearer <token>`.
+
+    Learn more about:
+        HTTP authentication schemes:
+            https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication
+        OAuth 2.0 Bearer authentication scheme:
+            https://oauth.net/2/bearer-tokens/
+        OpenID Connect 1.0 ID Tokens:
+            https://openid.net/specs/openid-connect-core-1_0.html#IDToken
+
+    Args:
+        request: webapp2.Request. The HTTP request to inspect.
+
+    Returns:
+        str|None. The ID Token of the request, if present, otherwise None.
+    """
+    scheme, _, token = request.headers.get('Authorization', '').partition(' ')
+    return token if scheme == 'Bearer' else None
+
+
+def _get_auth_claims_from_session_cookie(cookie):
+    """Returns claims from the session cookie, or None if invalid.
+
+    Args:
+        cookie: str|None. The session cookie to extract claims from.
```

Comment:
Seems weird, why allow None as a valid input?

(I don't feel too strongly though. But None is, strictly speaking, not a cookie.)

Reply:
Because it isn't an error for session cookies to be None, and handling it in this function makes surrounding code cleaner.